pkg/operator/target_config_reconciler.go

```diff
@@ -0,0 +1,274 @@
+	apiextensionsv1 "k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/v1"
+	apiextensionsclient "k8s.io/apiextensions-apiserver/pkg/client/clientset/clientset"
+	"k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/client-go/kubernetes"
+	"k8s.io/client-go/listers/core/v1"
+	"k8s.io/klog/v2"
+	"k8s.io/utils/ptr"
+
+	operatorv1 "github.com/openshift/api/operator/v1"
+	"github.com/openshift/library-go/pkg/controller/factory"
+	"github.com/openshift/library-go/pkg/operator/events"
+	"github.com/openshift/library-go/pkg/operator/resource/resourceapply"
+	"github.com/openshift/library-go/pkg/operator/resource/resourcemerge"
+	"github.com/openshift/library-go/pkg/operator/resource/resourceread"
+	"github.com/openshift/library-go/pkg/operator/v1helpers"
+
+	"github.com/openshift/jobset-operator/bindata"
+	jobsetoperatorsv1 "github.com/openshift/jobset-operator/pkg/apis/openshiftoperator/v1"
+	operatorclientinformers "github.com/openshift/jobset-operator/pkg/generated/informers/externalversions/openshiftoperator/v1"
+	openshiftoperatorv1 "github.com/openshift/jobset-operator/pkg/generated/listers/openshiftoperator/v1"
+	"github.com/openshift/jobset-operator/pkg/operator/operatorclient"
+)
+
+type TargetConfigReconciler struct {
+	targetImagePullSpec string
+	operatorNamespace   string
+
+	jobSetOperatorClient       *operatorclient.JobSetOperatorClient
+	kubeClient                 kubernetes.Interface
+	apiextensionsClient        *apiextensionsclient.Clientset
+	eventRecorder              events.Recorder
+	kubeInformersForNamespaces v1helpers.KubeInformersForNamespaces
+	jobSetOperatorsLister      openshiftoperatorv1.JobSetOperatorLister
+	secretsLister              v1.SecretLister
+	resourceCache              resourceapply.ResourceCache
+}
+
+func NewTargetConfigReconciler(targetImagePullSpec, operatorNamespace string,
+	operatorClientInformer operatorclientinformers.JobSetOperatorInformer,
+	kubeInformersForNamespaces v1helpers.KubeInformersForNamespaces,
+	jobSetOperatorClient *operatorclient.JobSetOperatorClient,
+	kubeClient kubernetes.Interface,
+	apiextensionsClient *apiextensionsclient.Clientset,
+	eventRecorder events.Recorder,
+) factory.Controller {
+	t := &TargetConfigReconciler{
+		targetImagePullSpec:   targetImagePullSpec,
+		operatorNamespace:     operatorNamespace,
+		jobSetOperatorClient:  jobSetOperatorClient,
+		kubeClient:            kubeClient,
+		apiextensionsClient:   apiextensionsClient,
+		secretsLister:         kubeInformersForNamespaces.SecretLister(),
+		jobSetOperatorsLister: operatorClientInformer.Lister(),
+		eventRecorder:         eventRecorder,
+		resourceCache:         resourceapply.NewResourceCache(),
+	}
+	return factory.New().WithInformers(
+		// for the operator changes
+		operatorClientInformer.Informer(),
+		// for the deployment and its configmap and secret
+		kubeInformersForNamespaces.InformersFor(operatorNamespace).Apps().V1().Deployments().Informer(),
```

Comment:
We talked this on Slack and I'm wondering should we also listen `service`, `servicemonitor`, even `serviceaccount` to trigger the reconciliation if these resources are modified as well?. 

On the other hand, listening every resource would be non-performant and overkill.

Reply:
IMO, we do not have to have informers for resources that we do not expect to change during the lifetime of the deployment, so configmap and secrets should be enough

Reply:
If the other resources are removed somehow, we will still get a periodic sync for them

Reply:
Yes. You are right. Period sync is a nice fall back mechanism to cover all the rest.